File: javaclient/src/main/java/jagex2/dash3d/AnimBase.java
package jagex2.dash3d;

import deob.ObfuscatedName;
import jagex2.io.Packet;

@ObfuscatedName("g")
public class AnimBase {

	@ObfuscatedName("g.b")
	public int size;

	@ObfuscatedName("g.c")
	public int[] types;

	@ObfuscatedName("g.d")
	public int[][] labels;

	public AnimBase(Packet buf) {
		this.size = buf.g1();

		this.types = new int[this.size];
		this.labels = new int[this.size][];

		for (int i = 0; i < this.size; i++) {
			this.types[i] = buf.g1();
		}

		for (int i = 0; i < this.size; i++) {
			int count = buf.g1();

			this.labels[i] = new int[count];
			for (int j = 0; j < count; j++) {
				this.labels[i][j] = buf.g1();
			}
		}
	}
}

File: engine/src/cache/graphics/AnimBase.ts
import Packet from '#/io/Packet.js';

export default class AnimBase {
    static instances: AnimBase[] = [];
    static order: string[] = [];

    static OP_BASE = 0;
    static OP_TRANSLATE = 1;
    static OP_ROTATE = 2;
    static OP_SCALE = 3;
    static OP_ALPHA = 5;

    length: number = 0;
    types: Int32Array = new Int32Array();
    labels: Int32Array[] = [];

    static unpack(dat: Packet): number {
        const length = dat.g1();

        const types = new Int32Array(length);
        const labels = new Array(length);

        for (let i = 0; i < length; i++) {
            types[i] = dat.g1();
        }

        for (let i = 0; i < length; i++) {
            const labelCount = dat.g1();
            labels[i] = new Int32Array(labelCount);

            for (let j = 0; j < labelCount; j++) {
                labels[i][j] = dat.g1();
            }
        }

        const base = new AnimBase();
        base.length = length;
        base.types = types;
        base.labels = labels;
        return AnimBase.instances.push(base) - 1;
    }
}
